### ISSUE TITLE
Loading is suspended in the end of the first epoch by using pytorch.DALIGenericIterator and ops.ExternalSource

### ISSUE THREAD

**Issue by Beanocean** [NVIDIA/DALI#1134]
I am trying to implement a DALI loader which can shuffle the examples after every epoch in the training stage.

My iterator for traversing the whole dataset and feeding examples into the `ops.ExternalSource()` throws a `StopIteration` error when the file list is exhausted. See the code example in the bottom.

In my implementation, the DALIGenericIterator will get stuck at the end of the first epoch.

How should I implement my `FileListReader`?

``` python
#!/usr/bin/env python
# coding: utf-8

from __future__ import print_function
from random import shuffle
from nvidia.dali.pipeline import Pipeline
import nvidia.dali.ops as ops
import nvidia.dali.types as types
from nvidia.dali.plugin.pytorch import DALIGenericIterator
import os
import numpy as np
from .utils import load_and_check_flist

class FileListReader(object):
    """ produce the files according to a given file list iteratively.  """
    def __init__(self, data_root, file_list, batch_size, split='train'):
        self.data_dir = data_root
        self.batch_size = batch_size
        self.data_list = load_and_check_flist(split, data_root, file_list)

        self.i = 0
        self.n = len(self.data_list)

        shuffle(self.data_list)  # TODO: shuffle here ??

    @property
    def size(self,):
        return self.n

    def __iter__(self,):
        self.i = 0
        self.n = len(self.data_list)
        return self

    def __next__(self):
        try:
            batch = []
            labels = []
            for _ in range(self.batch_size):
                source_path, target_path = self.data_list[self.i]
                print("===> {} / {}".format(self.i, self.n))

                # batch.append(np.frombuffer(open(source_path, 'rb').read(), dtype=np.uint8))
                # labels.append(np.frombuffer(open(target_path, 'rb').read(), dtype=np.uint8))

                # reading without numpy transformation is also ok
                batch.append(open(source_path, 'rb').read())
                labels.append(open(target_path, 'rb').read())

                self.i = (self.i + 1)

            return (batch, labels,)
        except:
            self.i = 0
            shuffle(self.data_list)  # TODO: shuffle here ??
            raise StopIteration

    next = __next__


class MyPipeline(Pipeline):
    def __init__(
        self, data_root, file_list, batch_size,
        num_threads=4, device_id=1, split='train'
    ):
        super(MyPipeline, self).__init__(
            batch_size, num_threads, device_id, seed=12)

        self.dataset = FileListReader(
            data_root, file_list, batch_size, split
        )

        self.source_feeder = ops.ExternalSource()
        self.target_feeder = ops.ExternalSource()

        self.source_decoder = ops.ImageDecoder(
            device='mixed', output_type=types.RGB)
        self.target_decoder = ops.ImageDecoder(
            device='mixed', output_type=types.GRAY)

        self.cast = ops.Cast(device='gpu', dtype=types.UINT8)

        self.resize = ops.Resize(device = "gpu", resize_shorter = 240)

    def define_graph(self,):
        self.source = self.source_feeder()
        self.target = self.target_feeder()
        image = self.source_decoder(self.source)
        label = self.target_decoder(self.target)

        image = self.resize(image)
        label = self.resize(label)

        return self.cast(image), self.cast(label)

    def iter_setup(self,):
        print(self.dataset.i, self.dataset.n)
        images, labels = self.dataset.next()
        self.feed_input(self.source, images)
        self.feed_input(self.target, labels)

    @property
    def size(self,):
        return self.dataset.size


def get_loader(split='train', batch_size=1, device_id=0):
    pipe = MyPipeline(
        'data/cityscapes', 'data/list/example.lst',
        batch_size, 4, device_id, 'train')
    pipe.build()
    return DALIGenericIterator(pipe, ['images', 'labels'], size=pipe.size)

def _test():
    loader = get_loader()
    for e in range(3):
        print('epoch:', e)
        for i, outputs in enumerate(loader):
            print(">>", i)
            images = outputs[0]['images'].cpu()
            labels = outputs[0]['labels'].cpu().squeeze()
            print(images.shape)
            print(labels.shape)

        print('reset loader')
        loader.reset()

if __name__ == "__main__":
    _test()
```

**Comment by awolant** [comment]
Hi, thanks for the question.
As you can see in the [ExternalSource example](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/external_input.html) iterator should work continuously and not raise the `StopIteration`. What you can do is to track when the file list is exhausted and then do the shuffling and continue returning samples.
Actually, DALI [FileReader op](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/external_input.html) already supports `shuffle_after_epoch` parameter, so maybe you can use that.

**Comment by Beanocean** [comment]
Albert @awolant , Thanks a lot for your reply,

Actually, I have read the code example of ExternalSource and noticed the iterator is endless.

In my situation, I want to implement the iterator to support shuffle_after_epoch and the DALIGenericIterator to reach the StopIteration in every epoch. Besides, the FileReader op only supports returning the numeric labels, I want to use an image as the label.

**Comment by JanuszL** [comment]
@Beanocean - in such case you need to reinitialize your `self.dataset` iterator in the `iter_setup`. It should be possible to raise `StopIteration` form the `iter_setup` but it doesn't work now and I just fixed it in https://github.com/NVIDIA/DALI/pull/1136. But even after the fix you still need to reinitialize your iterator).
So until the fix is merged you should do:
```
 def iter_setup(self,):,
        try:
            print(self.dataset.i, self.dataset.n)
            images, labels = self.dataset.next()
            self.feed_input(self.source, images)
            self.feed_input(self.target, labels)
    except StopIteration:
           images, labels = self.dataset.next()
           self.dataset = FileListReader(self.data_root, file_list, self.batch_size, self.split)
           self.feed_input(self.source, images)
           self.feed_input(self.target, labels)
```
After the fix it should be possible:
```
 def iter_setup(self,):,
        try:
            print(self.dataset.i, self.dataset.n)
            images, labels = self.dataset.next()
            self.feed_input(self.source, images)
            self.feed_input(self.target, labels)
    except StopIteration:
           self.dataset = FileListReader(self.data_root, file_list, self.batch_size, self.split)
           raise StopIteration
```

**Comment by Beanocean** [comment]
@JanuszL Thanks for your prompt response.

One more question, In the example where the bug is not fixed.
``` python
def iter_setup(self,):,
    try:
        print(self.dataset.i, self.dataset.n)
        images, labels = self.dataset.next()
        self.feed_input(self.source, images)
        self.feed_input(self.target, labels)
    except StopIteration:
        images, labels = self.dataset.next()
        self.dataset = FileListReader(self.data_root, file_list, self.batch_size, self.split)
        self.feed_input(self.source, images)
        self.feed_input(self.target, labels)
```
 why should we call the `images, labels = self.dataset.next()` after catching the **StopIteratin** error?

**Comment by JanuszL** [comment]
My bad, it should be:
```
def iter_setup(self,):,
    try:
        print(self.dataset.i, self.dataset.n)
        images, labels = self.dataset.next()
        self.feed_input(self.source, images)
        self.feed_input(self.target, labels)
    except StopIteration:
        self.dataset = FileListReader(self.data_root, file_list, self.batch_size, self.split)
        images, labels = self.dataset.next()
        self.feed_input(self.source, images)
        self.feed_input(self.target, labels)
```
You need to feed you source even if `StopIteration` is raised or your pipeline will run out of data and (due to bug #1136) iterator is not able to refill it again.

**Comment by Beanocean** [comment]
@JanuszL Cool! I've already used this code clip in my project. Thanks again.

**Comment by justinblaber** [comment]
This looks similar to #1604. One problem with the above solution is it will skip the last batch if its smaller than the batch size... although it probably doesnt matter since the data is shuffled every epoch.

**Comment by JanuszL** [comment]
You can provide dummy data to fill the rest of the batch and then use `last_batch_padded` in the framework iterator. When you use the bare DALI pipeline then you can just wrap up and use the data from the next epoch to fill the remaining samples.
